File: desktop_editor/DynamicDriving.Core/Playback/OverlayGroupScheduler.cs
using DynamicDriving.Core.Manifests;

namespace DynamicDriving.Core.Playback;

public readonly record struct OverlayGroupState(
    string? ActiveStemId = null,
    long ActiveUntilMs = 0,
    long CooldownUntilMs = 0);

public abstract record OverlayGroupAction
{
    public sealed record None : OverlayGroupAction;

    public sealed record Start(StemManifest Stem, long ActiveUntilMs) : OverlayGroupAction;

    public sealed record Stop(string StemId, long CooldownUntilMs) : OverlayGroupAction;
}

public readonly record struct OverlayGroupDecision(OverlayGroupState State, OverlayGroupAction Action);

/// <summary>
/// One overlay per group at a time, with a randomised cooldown between plays.
///
/// This is the same scheduler the Android engine runs, so the editor's preview shows the real
/// pacing rather than firing every eligible overlay at once.
/// </summary>
public static class OverlayGroupScheduler
{
    public static OverlayGroupDecision Tick(
        OverlayGroupState currentState,
        long nowMs,
        IReadOnlyList<StemManifest> eligibleCandidates,
        StemManifest? activeStem,
        bool forceStopActive,
        Random random)
    {
        if (currentState.ActiveStemId is not null && activeStem is not null)
        {
            var overlayRule = (OverlayStemRule)activeStem.Rule;
            if (forceStopActive || nowMs >= currentState.ActiveUntilMs)
            {
                var cooldownUntilMs = nowMs + PlaybackRules.NextCooldownMs(overlayRule, random);
                return new OverlayGroupDecision(
                    currentState with
                    {
                        ActiveStemId = null,
                        ActiveUntilMs = 0,
                        CooldownUntilMs = cooldownUntilMs
                    },
                    new OverlayGroupAction.Stop(activeStem.StemId, cooldownUntilMs));
            }

            return new OverlayGroupDecision(currentState, new OverlayGroupAction.None());
        }

        if (nowMs < currentState.CooldownUntilMs)
        {
            return new OverlayGroupDecision(currentState, new OverlayGroupAction.None());
        }

        var selectedStem = PlaybackRules.ChooseWeightedOverlay(eligibleCandidates, random);
        if (selectedStem is null)
        {
            return new OverlayGroupDecision(currentState, new OverlayGroupAction.None());
        }

        var selectedRule = (OverlayStemRule)selectedStem.Rule;
        var activeUntilMs = nowMs + selectedRule.DurationMs;
        return new OverlayGroupDecision(
            new OverlayGroupState(selectedStem.StemId, activeUntilMs, currentState.CooldownUntilMs),
            new OverlayGroupAction.Start(selectedStem, activeUntilMs));
    }
}

File: shared/src/main/java/com/BalancedLight/dynamicdriving/shared/playback/OverlayGroupScheduler.kt
package com.BalancedLight.dynamicdriving.shared.playback

import com.BalancedLight.dynamicdriving.shared.catalog.OverlayStemRule
import com.BalancedLight.dynamicdriving.shared.catalog.StemManifest
import kotlin.random.Random

data class OverlayGroupState(
    val activeStemId: String? = null,
    val activeUntilMs: Long = 0L,
    val cooldownUntilMs: Long = 0L
)

sealed interface OverlayGroupAction {
    data object None : OverlayGroupAction
    data class Start(val stem: StemManifest, val activeUntilMs: Long) : OverlayGroupAction
    data class Stop(val stemId: String, val cooldownUntilMs: Long) : OverlayGroupAction
}

data class OverlayGroupDecision(
    val state: OverlayGroupState,
    val action: OverlayGroupAction
)

object OverlayGroupScheduler {
    fun tick(
        currentState: OverlayGroupState,
        nowMs: Long,
        eligibleCandidates: List<StemManifest>,
        activeStem: StemManifest?,
        forceStopActive: Boolean,
        random: Random
    ): OverlayGroupDecision {
        if (currentState.activeStemId != null && activeStem != null) {
            val overlayRule = activeStem.rule as OverlayStemRule
            if (forceStopActive || nowMs >= currentState.activeUntilMs) {
                val cooldownUntilMs = nowMs + PlaybackRules.nextCooldownMs(overlayRule, random)
                return OverlayGroupDecision(
                    state = currentState.copy(
                        activeStemId = null,
                        activeUntilMs = 0L,
                        cooldownUntilMs = cooldownUntilMs
                    ),
                    action = OverlayGroupAction.Stop(activeStem.stemId, cooldownUntilMs)
                )
            }
            return OverlayGroupDecision(currentState, OverlayGroupAction.None)
        }

        if (nowMs < currentState.cooldownUntilMs) {
            return OverlayGroupDecision(currentState, OverlayGroupAction.None)
        }

        val selectedStem = PlaybackRules.chooseWeightedOverlay(eligibleCandidates, random)
            ?: return OverlayGroupDecision(currentState, OverlayGroupAction.None)
        val overlayRule = selectedStem.rule as OverlayStemRule
        val activeUntilMs = nowMs + overlayRule.durationMs
        return OverlayGroupDecision(
            state = OverlayGroupState(
                activeStemId = selectedStem.stemId,
                activeUntilMs = activeUntilMs,
                cooldownUntilMs = currentState.cooldownUntilMs
            ),
            action = OverlayGroupAction.Start(selectedStem, activeUntilMs)
        )
    }
}
